### PR TITLE
Frontend changes to watch VolumeSnapshot CR

### PR DESCRIPTION
**Pull request by pavannd1**

### CODE REVIEW

File: frontend/kubernetes/plugin.go [hunk 1]
```diff
@@ -1508,3 +1542,178 @@ func (p *Plugin) updatePVCSize(
...
+	p.processAddedVolumeSnapshot(volumeSnapshot)
+}
+
+func (p *Plugin) processAddedVolumeSnapshot(volSnap *k8ssnapshotv1alpha1.VolumeSnapshot) {
+
+	// Validate the VolumeSnapshot CR
+	if volSnap.Status.Error != nil {
+		log.WithFields(log.Fields{
+			"volumeSnapshot": volSnap.Name,
+		}).Error("Kubernetes frontend found error set on the VolumeSnapshot")
```
Comment (tdmanv) [inline frontend/kubernetes/plugin.go]: This error would have already been logged. I would just return.

File: frontend/kubernetes/plugin.go [hunk 2]
```diff
@@ -1508,3 +1542,178 @@ func (p *Plugin) updatePVCSize(
...
+	if err != nil {
+		log.WithFields(log.Fields{
+			"namespace": volSnap.Namespace,
+			"error":     err.Error(),
+		}).Error("Kubernetes frontend couldn't create a client to namespace!")
+		return
+	}
+	claim, err := k8sClient.GetPVC(sourcePVC, metav1.GetOptions{})
+	if err != nil {
+		err = fmt.Errorf("snapshotting a PVC requires the PVC to be in the same namespace as the VolumeSnapshot CR")
```
Comment (tdmanv) [inline frontend/kubernetes/plugin.go]: did you mean to overwrite the error here?

File: frontend/kubernetes/plugin.go [hunk 3]
```diff
@@ -1508,3 +1542,178 @@ func (p *Plugin) updatePVCSize(
...
+			"sourcePVC":      sourcePVC,
+		}).Errorf("Kubernetes frontend detected an invalid configuration "+
+			"for creating snapshot from a PVC: %v", err.Error())
+		return
+	}
+
+	// 2. Trident should be set as the provisioner for the PVC
+	provisioner := getClaimProvisioner(claim)
+	if provisioner != AnnOrchestrator &&
+		GetPersistentVolumeClaimClass(claim) != "" {
```
Comment (tdmanv) [inline frontend/kubernetes/plugin.go:1623]: Please excuse my ignorance. What does this check do?

File: frontend/kubernetes/plugin.go [hunk 4]
```diff
@@ -1508,3 +1542,178 @@ func (p *Plugin) updatePVCSize(
...
+			"sourcePVC":       claim.Name,
+			"sourcePVCStatus": claim.Status.Phase,
+		}).Debug("Kubernetes frontend ignored this VolumeSnapshot as the source PVC is not " +
+			"bound to any volume")
+		return
+	}
+
+	// Create the volume configuration object
+	uniqueClaimName := getUniqueClaimName(claim)
+	size, _ := claim.Spec.Resources.Requests[v1.ResourceStorage]
```
Comment (tdmanv) [inline frontend/kubernetes/plugin.go]: does this need to be in multi-value context?

File: frontend/kubernetes/plugin.go [hunk 5]
```diff
@@ -1508,3 +1542,178 @@ func (p *Plugin) updatePVCSize(
...
+	// Create the snapshot
+	snapshotExt, err := p.orchestrator.CreateVolumeSnapshot(volSnap.Name, volConfig)
+	if err != nil {
+		newVolSnap, uerr := p.updateVolumeSnapshotErrorStatus(volSnap, err.Error())
+		if uerr != nil {
+			p.eventRecorder.Event(volSnap, v1.EventTypeWarning, "UpdateStatusFailed",
+				"Snapshots controller failed to update VolumeSnapshotErrorStatus")
+			return
+		}
+		p.eventRecorder.Event(newVolSnap, v1.EventTypeNormal, "CreateSnapshotFailed", err.Error())
```
Comment (tdmanv) [inline frontend/kubernetes/plugin.go]: Is this a normal event? why not make this a warning or error?

File: frontend/kubernetes/plugin.go [hunk 6]
```diff
@@ -1508,3 +1542,178 @@ func (p *Plugin) updatePVCSize(
...
+		return
+	}
+
+	// Parse creation time
+	var createdAt time.Time
+	if createdAt, err = time.Parse(time.RFC3339, snapshotExt.Snapshot.Created); err != nil {
```
Comment (tdmanv) [inline frontend/kubernetes/plugin.go]: [minor] why not push this down into updateVolumeSnapshotStatus. You can then just pass `volSnap` and `snapshotExt` as parameters.

File: frontend/kubernetes/plugin.go [hunk 7]
```diff
@@ -244,6 +258,24 @@ func newKubernetesPlugin(
...
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			return kubeSnapClient.VolumesnapshotV1alpha1().VolumeSnapshots(v1.NamespaceAll).Watch(options)
+		},
+	}
+	_, ret.volumeSnapshotController = cache.NewInformer(
+		ret.volumeSnapshotSource,
+		&k8ssnapshotv1alpha1.VolumeSnapshot{},
+		KubernetesSyncPeriod,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: ret.addVolumeSnapshot,
```
Comment (tdmanv) [inline frontend/kubernetes/plugin.go]: This should also handle update correctly - Might as well also set this callback for the handler.

File: frontend/kubernetes/plugin.go (hunk 3 above)
Comment (pavannd1) [reply frontend/kubernetes/plugin.go:1623]: This is checking if the source PVC was provisioned by Trident.

File: frontend/kubernetes/plugin.go (hunk 2 above)
Comment (pavannd1) [reply frontend/kubernetes/plugin.go]: Yes, this was intentional. Just sticking to their pattern.

**Review by tdmanv — APPROVED** [review]

**Pull request closed without merging**